Sources/Operator/py_local_operator.hpp

```diff
@@ -126,6 +126,10 @@ void AddLocalOperator(py::module &subm) {
       .def_property_readonly(
           "acting_on", &LocalOperator::ActingOn,
           R"EOF(list[list]: A list of the sites that each local matrix acts on.)EOF")
+      .def("transpose", &LocalOperator::Transpose,
+           R"EOF(Returns the transpose of this operator)EOF")
+      .def("conjugate", &LocalOperator::Conjugate,
```

Comment:
@gcarleo please verify if those bindings are declared correctly. I am unsure if I should add a `py::keep_alive` declaration. (I think that I should, because I am creating a new object)

Reply:
here you are calling the constructor of `LocalOperator` and returning the object it creates. I think you don't need to keep (hilbert?) alive explicitly, since this is already done by the constructor. @twesterhout @femtobit do you agree?

Reply:
This indeed causes a memory management issue. The following code causes a segfault (using CPython):
```python
import netket as nk

g = nk.graph.Hypercube(8, 1)
hi = nk.hilbert.Spin(g, 1/2)

lo = nk.operator.LocalOperator(hi, [[1,0],[0,1]], [0])
lo = lo.transpose()

# Assign hi to something else so it will no longer keep alive the original Hilbert object
hi = nk.hilbert.Qubit(g)

# This will try to access the now deleted Hilbert object
lo = lo * lo
```

There are two reasonable solutions:
1. Remove the reference to Hilbert in the operator (it is not really used, except for `GetHilbert`).
2. Store the `AbstractHilbert` as an `std::shared_ptr`. This requires changes in several places in the codebase, so implementing this option should probably wait a few days until @twesterhout is done with his current restructuring of the code.

Reply:
@femtobit Thanks for pointing that out. But this bug is not introduced by my PR, I'm hitting an existing bug in netket (just replace `lo = lo.transpose()` above with `lo=lo*lo`).

Regarding option 1: I ran a quick search and apparently `GetHilbert` is used in the initialisation of Matrix Wrappers and somewhere in `CheckMoveOperators` and `Init` of `custom_sampler`. Do you think it would be reasonable to remove it? (Besides, I don't think this is a smart move)

I would more lean towards option 2: - I already believe that netket uses references too heavily, where we should be instead using `std::shared_ptr`. This is one of such cases. Do you agree?

Reply:
Indeed this is not a PR-specific issue, thus we can merge it for now, and then this will hopefully be fixed once we handle `Hilbert` properly. 
I also agree that option 1 in here is not viable, because an operator **must** be able to return the Hilbert space it is acting on! 

Again, my favorite solution would be to make these types concrete, but I need to think more about the consequences and how to do that nicely